packages/SwingSet/src/kernel/vatManager/supervisor-subprocess-xsnap.js

```diff
@@ -0,0 +1,237 @@
+// @ts-check
+import { assert, details } from '@agoric/assert';
+import { importBundle } from '@agoric/import-bundle';
+import { Remotable, getInterfaceOf, makeMarshal } from '@agoric/marshal';
+// grumble... waitUntilQuiescent is exported and closes over ambient authority
+import { waitUntilQuiescent } from '../../waitUntilQuiescent';
+
+import { makeLiveSlots } from '../liveSlots';
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+// eslint-disable-next-line no-unused-vars
+function workerLog(first, ...args) {
+  // @ts-ignore
+  // eslint-disable-next-line
+  // print(`---worker: ${first}`, ...args);
+}
+
+workerLog(`supervisor started`);
+
+/**
+ * @typedef { [unknown, ...unknown[]] } Tagged
+ */
+const Item = {
```

Comment:
I think there’s an analogy here to `TextEncoder` and `TextDecoder`. It those were combined into one object, they’d be called `TextCodec`. An instance of that would be `textCodec`, or just `codec` if that’s unambiguous in scope. Since this is a singleton, I’d suggest instance naming conventions rather than class naming.

This might be something to capture in a separate module since it might see reuse. If we did some xsnap refactoring (reorienting around Uint8Array instead of ArrayBuffer), it might be portable over to Node.js.

I’d also like to keep in mind that this is the layer where we will have an opportunity to mux binary and JSON messages. This is where we would add support for marshaling typed arrays over CapTP. To do that, we might need another thing layer of enveloping as the protocol evolves, e.g., a `\0` prefix instead of `[` to distinguish a message containing just a chunk of bytes that can be referenced in a subsequent dispatch by its sequence id.